10-JavaScript/03-objects.md

```diff
@@ -0,0 +1,133 @@
+
+```javascript
+var myDog = {
+  name: 'fido',
+  breed: 'labrador',
+  age: 4,
+  speak: function() {
+    console.log('woof');
+  },
+  owner: {
+    name: 'Ada',
+    address: '1215 4th Ave #1050'
+  },
+  toString: function() {
+    return this.name + ", a " + this.breed " owned by " + this.owner.name;
+  }
+};
+```
+
+This is great for one-off objects, but what if we need to keep track of many similar objects? We'd have to explicitly define each one, which could quickly turn into a lot of repeated code.
+
+### Instantiating Objects
+
+#### Ruby-style: Classical Objects
+In Ruby, we solve the problem of many similar objects using classes. We define a class, give it some methods and variables, and then create and manipulate instances of the class.
+
+```ruby
+# dog.rb
+class Dog
+  attr_reader :name, :breed
+  def initialize(name, breed)
+    @name = name
+    @breed = breed
+  end
+
+  def speak()
+    puts "woof"
+  end
+end
+
+myDog = Dog.new('fido', 'labrador')
+myDog.speak
+```
+
+This pattern of defining a class and then instantiating it is very common in object oriented languages, and languages that use it are sometimes said to have _classical_ objects. JavaScript uses a slightly different technique called _prototypical_ objects.
+
+#### JavaScript Style: Prototypical Objects
+When building a prototypical object in JavaScript, there are three important steps:
+- Define the _constructor_, a function that sets up any details specific to this instance (usually instance variables)
+- Define the constructor's `prototype`, an object that establishes pieces common to all instances (usually instance methods)
+- Call the constructor using JavaScript's `new` keyword to create new instances of the object
+
+The big idea of prototypical objects is that, rather than defining a special language construct like a class, you create an original object to be the _prototype_ for new objects. Any new object is created as a clone of the original.
+
+```javascript
+// dog.js
+
+// First we define the constructor
+var Dog = function(name, breed) {
+  this.name = name;
+  this.breed = breed;
+};
+
+// Second we add the prototype
+Dog.prototype = {
+  speak: function () {
+    console.log("woof")
+  }
+};
+
+// Third we instantiate the object with the new keyword
+var myDog = new Dog("fido", "labrador");
+myDog.speak();
+```
+
+Now we can create as many dogs as we want!
+
+Note the `new` keyword. The syntax is a little different from Ruby, and it's easy to get tripped up.
+
+Note also that `Dog.prototype` is just a JavaScript object. It happens to be the object that JavaScript looks for when a constructor is invoked with `new`, but other than that it behaves just as we've seen before. So you could do things like:
+
+```javascript
+// Call a function from the prototype directly
+Dog.prototype.speak();
+
+// Add a function to the prototype
+Dog.prototype.isHungry = function() {
```

Comment:
It might be worthwhile to continue this example of extending the prototype, but I think it would depend on the class. It doesn't seem like it is a major learning goal.